sdk/storage/azure-storage-queue/src/main/java/com/azure/storage/queue/QueueClient.java

```diff
@@ -948,7 +1258,92 @@ public PeekedMessageItem peekMessage() {
     public PagedIterable<PeekedMessageItem> peekMessages(Integer maxMessages, Duration timeout, Context context) {
-        return new PagedIterable<>(client.peekMessagesWithOptionalTimeout(maxMessages, timeout, context));
+        return peekMessagesWithOptionalTimeout(maxMessages, timeout, context);
+    }
+
+    PagedIterable<PeekedMessageItem> peekMessagesWithOptionalTimeout(Integer maxMessages, Duration timeout, Context context) {
+        Context finalContext = context == null ? Context.NONE : context;
+        try {
+            Supplier<ResponseBase<MessagesPeekHeaders, List<PeekedMessageItemInternal>>> operation = () ->
+                this.azureQueueStorage.getMessages().peekWithResponse(queueName, maxMessages, null, null, finalContext);
+            ResponseBase<MessagesPeekHeaders, List<PeekedMessageItemInternal>> response = timeout != null
+                ? THREAD_POOL.submit(operation::get).get(timeout.toMillis(), TimeUnit.MILLISECONDS) : operation.get();
+
+            PagedResponseBase<MessagesPeekHeaders, PeekedMessageItem> transformedMessages =
+                transformMessagesPeekResponse(response);
+
+            Supplier<PagedResponse<PeekedMessageItem>> res = () -> new PagedResponseBase<>(
+                response.getRequest(),
+                response.getStatusCode(),
+                response.getHeaders(),
+                transformedMessages,
+                response.getDeserializedHeaders());
+
+            return new PagedIterable<>(res);
+
+        } catch (RuntimeException e) {
+            throw LOGGER.logExceptionAsError(e);
+        } catch (InterruptedException | ExecutionException | TimeoutException e) {
+            throw LOGGER.logExceptionAsError(new RuntimeException(e));
+        }
+    }
+
+    private PagedResponseBase<MessagesPeekHeaders, PeekedMessageItem> transformMessagesPeekResponse(
+        ResponseBase<MessagesPeekHeaders, List<PeekedMessageItemInternal>> response) {
+        List<PeekedMessageItemInternal> peekedMessageInternalItems = response.getValue();
+        if (peekedMessageInternalItems == null) {
+            peekedMessageInternalItems = Collections.emptyList();
+        }
+        List<PeekedMessageItem> messageItems = new ArrayList<>();
+        for (PeekedMessageItemInternal peekedMessageInternalItem : peekedMessageInternalItems) {
+            try {
+                PeekedMessageItem peekedMessageItem =
+                    transformPeekedMessageItemInternal(peekedMessageInternalItem, messageEncoding);
+                messageItems.add(peekedMessageItem);
+
+            } catch (IllegalArgumentException e) {
+                if (processMessageDecodingErrorAsyncHandler != null) {
+                    PeekedMessageItem transformedPeekedMessageItem =
+                        transformPeekedMessageItemInternal(peekedMessageInternalItem, QueueMessageEncoding.NONE);
+
+                    processMessageDecodingErrorAsyncHandler.apply(new QueueMessageDecodingError(
```

Comment:
Looks like there's some duplication in all the error handling scenarios. Can we extract this out to a common method?

Reply:
Not exactly sure how we can extract out the logic, since receiveMessages() transforms and returns a QueueMessageItem, whereas peekMessages() transforms and returns a PeekedMessageItem. The logic is also similar in the async client, but it's all dealing with async calls.